File: lib/target-db.ts
type TargetDbResult = {
  ok: boolean;
  rowsAffected?: number;
  error?: string;
  sql?: string;
};

type TargetRowsResult = {
  ok: boolean;
  rows: Record<string, unknown>[];
  columns: string[];
  error?: string;
};

type IbmDbModule = {
  open: (connectionString: string, callback: (error: Error | null, conn: any) => void) => void;
};

function targetDb() {
  return (process.env.TARGET_DB || "postgres").toLowerCase();
}

function quoteIdentifier(identifier: string) {
  const parts = identifier.split(".");
  if (!parts.every((part) => /^[A-Z_][A-Z0-9_]*$/i.test(part))) {
    throw new Error(`Identifiant SQL non autorise : ${identifier}`);
  }
  return parts.map((part) => part.toUpperCase()).join(".");
}

function targetTable() {
  return quoteIdentifier(process.env.TARGET_TABLE || process.env.DB2_TARGET_TABLE || "CRCON");
}

function targetKeyColumn() {
  return quoteIdentifier(process.env.TARGET_KEY_COLUMN || process.env.DB2_CONTRACT_COLUMN || "NODOSS");
}

export function getTargetInfo() {
  return {
    db: targetDb(),
    table: targetTable(),
    keyColumn: targetKeyColumn()
  };
}

export function buildTargetUpdatePreview(input: { updates: { dbColumn: string }[] }) {
  const db = targetDb();
  if (db === "bridge") return buildBridgeSql(input);
  if (db === "db2") return buildDb2Sql(input);
  return buildPostgresSql(input);
}

export function buildFilledTargetUpdatePreview(input: {
  contractNumber: string;
  updates: { dbColumn: string; value: string }[];
}) {
  const db = targetDb();
  if (db === "bridge") return buildFilledBridgeSql(input);
  if (db === "db2") return buildFilledDb2Sql(input);
  return buildFilledPostgresSql(input);
}

export async function updateTargetContract(input: {
  contractNumber: string;
  updates: { dbColumn: string; value: string }[];
}): Promise<TargetDbResult> {
  const db = targetDb();
  if (db === "db2") return updateDb2(input);
  if (db === "bridge") return updateBridge(input);
  return updatePostgres(input);
}

export async function readTargetRows(limit = 50): Promise<TargetRowsResult> {
  const db = targetDb();
  if (db === "db2") return readDb2Rows(limit);
  if (db === "bridge") return readBridgeRows(limit);
  return readPostgresRows(limit);
}

function buildDb2Sql(input: { updates: { dbColumn: string }[] }) {
  const assignments = input.updates.map((item) => `${quoteIdentifier(item.dbColumn)} = ?`).join(", ");
  return `UPDATE ${targetTable()}
SET ${assignments}, UPDATED_AT = CURRENT TIMESTAMP
WHERE ${targetKeyColumn()} = ?`;
}

function buildFilledDb2Sql(input: { contractNumber: string; updates: { dbColumn: string; value: string }[] }) {
  const assignments = input.updates
    .map((item) => `${quoteIdentifier(item.dbColumn)} = ${sqlLiteral(item.value)}`)
    .join(", ");
  return `UPDATE ${targetTable()}
SET ${assignments}, UPDATED_AT = CURRENT TIMESTAMP
WHERE ${targetKeyColumn()} = ${sqlLiteral(input.contractNumber)}`;
}

function buildBridgeSql(input: { updates: { dbColumn: string }[] }) {
  const assignments = input.updates.map((item) => `${quoteIdentifier(item.dbColumn)} = ?`).join(", ");
  return `UPDATE ${targetTable()}
SET ${assignments}
WHERE ${targetKeyColumn()} = ?`;
}

function buildFilledBridgeSql(input: { contractNumber: string; updates: { dbColumn: string; value: string }[] }) {
  const assignments = input.updates
    .map((item) => `${quoteIdentifier(item.dbColumn)} = ${sqlLiteral(item.value)}`)
    .join(", ");
  return `UPDATE ${targetTable()}
SET ${assignments}
WHERE ${targetKeyColumn()} = ${sqlLiteral(input.contractNumber)}`;
}

function buildPostgresSql(input: { updates: { dbColumn: string }[] }) {
  const assignments = input.updates
    .map((item, index) => `${quoteIdentifier(item.dbColumn)} = $${index + 1}`)
    .join(", ");
  return `UPDATE ${targetTable()}
SET ${assignments}, UPDATED_AT = CURRENT_TIMESTAMP
WHERE ${targetKeyColumn()} = $${input.updates.length + 1}`;
}

function buildFilledPostgresSql(input: { contractNumber: string; updates: { dbColumn: string; value: string }[] }) {
  const assignments = input.updates
    .map((item) => `${quoteIdentifier(item.dbColumn)} = ${sqlLiteral(item.value)}`)
    .join(", ");
  return `UPDATE ${targetTable()}
SET ${assignments}, UPDATED_AT = CURRENT_TIMESTAMP
WHERE ${targetKeyColumn()} = ${sqlLiteral(input.contractNumber)}`;
}

function sqlLiteral(value: string) {
  return `'${value.replaceAll("'", "''")}'`;
}

async function updatePostgres(input: {
  contractNumber: string;
  updates: { dbColumn: string; value: string }[];
}): Promise<TargetDbResult> {
  const connectionString = process.env.TARGET_DATABASE_URL || process.env.POSTGRES_URL;
  const sql = buildPostgresSql({ updates: input.updates });
  const params = [...input.updates.map((item) => item.value), input.contractNumber];

  if (!connectionString) {
    return { ok: false, error: "TARGET_DATABASE_URL n'est pas configure.", sql };
  }

  try {
    const { Client } = await import("pg");
    const client = new Client({ connectionString });
    await client.connect();
    const result = await client.query(sql, params);
    await client.end();
    const rowsAffected = result.rowCount ?? 0;
    return {
      ok: rowsAffected > 0,
      rowsAffected,
      error: rowsAffected === 0 ? "Postgres a mis a jour 0 ligne." : undefined,
      sql
    };
  } catch (error) {
    return { ok: false, error: error instanceof Error ? error.message : "Erreur Postgres inconnue.", sql };
  }
}

async function readPostgresRows(limit: number): Promise<TargetRowsResult> {
  const connectionString = process.env.TARGET_DATABASE_URL || process.env.POSTGRES_URL;
  const safeLimit = Math.max(1, Math.min(limit, 200));
  const sql = `SELECT * FROM ${targetTable()} ORDER BY ${targetKeyColumn()} LIMIT ${safeLimit}`;

  if (!connectionString) {
    return { ok: false, rows: [], columns: [], error: "TARGET_DATABASE_URL n'est pas configure." };
  }

  try {
    const { Client } = await import("pg");
    const client = new Client({ connectionString });
    await client.connect();
    const result = await client.query(sql);
    await client.end();
    const rows = result.rows.map((row) => normalizeRow(row));
    return { ok: true, rows, columns: collectColumns(rows) };
  } catch (error) {
    return {
      ok: false,
      rows: [],
      columns: [],
      error: error instanceof Error ? error.message : "Lecture Postgres impossible."
    };
  }
}

async function updateDb2(input: {
  contractNumber: string;
  updates: { dbColumn: string; value: string }[];
}): Promise<TargetDbResult> {
  const connectionString = process.env.DB2_CONNECTION_STRING;
  const sql = buildDb2Sql({ updates: input.updates });
  const params = [...input.updates.map((item) => item.value), input.contractNumber];

  if (!connectionString) {
    return { ok: false, error: "DB2_CONNECTION_STRING n'est pas configure.", sql };
  }

  try {
    const ibmdbModule = await import("ibm_db");
    const ibmdb = (ibmdbModule.default ?? ibmdbModule) as unknown as IbmDbModule;
    const connection = await new Promise<any>((resolve, reject) => {
      ibmdb.open(connectionString, (error: Error | null, conn: any) => {
        if (error) reject(error);
        else resolve(conn);
      });
    });

    try {
      const rowsAffected = await new Promise<number>((resolve, reject) => {
        connection.query(sql, params, (error: Error | null, _rows: unknown, info: { rowsAffected?: number }) => {
          if (error) reject(error);
          else resolve(info?.rowsAffected ?? 0);
        });
      });

      await new Promise<void>((resolve) => connection.close(() => resolve()));
      return {
        ok: rowsAffected > 0,
        rowsAffected,
        error: rowsAffected === 0 ? "DB2 a mis a jour 0 ligne." : undefined,
        sql
      };
    } catch (error) {
      await new Promise<void>((resolve) => connection.close(() => resolve()));
      throw error;
    }
  } catch (error) {
    return { ok: false, error: error instanceof Error ? error.message : "Erreur DB2 inconnue.", sql };
  }
}

async function readDb2Rows(limit: number): Promise<TargetRowsResult> {
  const connectionString = process.env.DB2_CONNECTION_STRING;
  const safeLimit = Math.max(1, Math.min(limit, 200));
  const sql = `SELECT * FROM ${targetTable()} ORDER BY ${targetKeyColumn()} FETCH FIRST ${safeLimit} ROWS ONLY`;

  if (!connectionString) {
    return { ok: false, rows: [], columns: [], error: "DB2_CONNECTION_STRING n'est pas configure." };
  }

  try {
    const ibmdbModule = await import("ibm_db");
    const ibmdb = (ibmdbModule.default ?? ibmdbModule) as unknown as IbmDbModule;
    const connection = await new Promise<any>((resolve, reject) => {
      ibmdb.open(connectionString, (error: Error | null, conn: any) => {
        if (error) reject(error);
        else resolve(conn);
      });
    });

    try {
      const dbRows = await new Promise<Record<string, unknown>[]>((resolve, reject) => {
        connection.query(sql, (error: Error | null, rows: Record<string, unknown>[]) => {
          if (error) reject(error);
          else resolve(rows);
        });
      });

      await new Promise<void>((resolve) => connection.close(() => resolve()));
      const rows = dbRows.map((row) => normalizeRow(row));
      return { ok: true, rows, columns: collectColumns(rows) };
    } catch (error) {
      await new Promise<void>((resolve) => connection.close(() => resolve()));
      throw error;
    }
  } catch (error) {
    return {
      ok: false,
      rows: [],
      columns: [],
      error: error instanceof Error ? error.message : "Lecture DB2 impossible."
    };
  }
}

function bridgeUrl() {
  return (process.env.BRIDGE_URL || "http://127.0.0.1:8001").replace(/\/$/, "");
}

async function updateBridge(input: {
  contractNumber: string;
  updates: { dbColumn: string; value: string }[];
}): Promise<TargetDbResult> {
  const sql = buildBridgeSql({ updates: input.updates });

  try {
    const response = await fetch(`${bridgeUrl()}/update`, {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify(input),
      cache: "no-store"
    });
    const payload = (await response.json().catch(() => ({}))) as {
      ok?: boolean;
      rowsAffected?: number;
      error?: string;
      detail?: string;
      sql?: string;
    };
    const error = payload.error || payload.detail;

    return {
      ok: response.ok && payload.ok === true,
      rowsAffected: payload.rowsAffected,
      error: response.ok ? error : error || `Bridge ODBC HTTP ${response.status}`,
      sql: payload.sql || sql
    };
  } catch (error) {
    return { ok: false, error: error instanceof Error ? error.message : "Bridge ODBC injoignable.", sql };
  }
}

async function readBridgeRows(limit: number): Promise<TargetRowsResult> {
  const safeLimit = Math.max(1, Math.min(limit, 200));

  try {
    const response = await fetch(`${bridgeUrl()}/rows?limit=${safeLimit}`, { cache: "no-store" });
    const payload = (await response.json().catch(() => ({}))) as {
      ok?: boolean;
      rows?: Record<string, unknown>[];
      columns?: string[];
      error?: string;
      detail?: string;
    };

    if (!response.ok || payload.ok !== true) {
      return {
        ok: false,
        rows: [],
        columns: [],
        error: payload.error || payload.detail || `Bridge ODBC HTTP ${response.status}`
      };
    }

    const rows = (payload.rows || []).map((row) => normalizeRow(row));
    return { ok: true, rows, columns: payload.columns || collectColumns(rows) };
  } catch (error) {
    return {
      ok: false,
      rows: [],
      columns: [],
      error: error instanceof Error ? error.message : "Bridge ODBC injoignable."
    };
  }
}

function normalizeRow(row: Record<string, unknown>) {
  return Object.fromEntries(Object.entries(row).map(([key, value]) => [key.toUpperCase(), value]));
}

function collectColumns(rows: Record<string, unknown>[]) {
  const preferred = [
    "NODOSS",
    "CDENVO",
    "NOCPA1",
    "NOCPA2",
    "NOCPA3",
    "NOCPA4",
    "NOCPA5",
    "CDAGEN",
    "NMTITU",
    "CINALP",
    "CINNUM",
    "CDMATR",
    "IMPUTA",
    "MTCRED",
    "MNTTOT",
    "TXTEG",
    "MTMENS",
    "NBMOIS",
    "DTECHD",
    "DTECHF",
    "UPDATED_AT"
  ];
  const found = new Set(rows.flatMap((row) => Object.keys(row)));
  return preferred.filter((column) => found.has(column)).concat([...found].filter((column) => !preferred.includes(column)));
}

File: workflow/target_db.py
from dataclasses import dataclass
import os
import re
import sqlite3
from pathlib import Path


IDENTIFIER_RE = re.compile(r"^[A-Z_][A-Z0-9_]*$", re.I)


@dataclass(frozen=True)
class TargetResult:
    ok: bool
    rows_affected: int = 0
    error: str = ""
    sql: str = ""


def quote_identifier(identifier: str) -> str:
    parts = identifier.split(".")
    if not all(IDENTIFIER_RE.match(part or "") for part in parts):
        raise ValueError(f"Identifiant SQL non autorise : {identifier}")
    return ".".join(part.upper() for part in parts)


def target_db() -> str:
    configured = os.environ.get("DJANGO_TARGET_DB", os.environ.get("TARGET_DB", "sqlite")).lower()
    return "sqlite" if configured == "bridge" else configured


def target_table() -> str:
    if target_db() == "sqlite":
        return quote_identifier(os.environ.get("TARGET_SQLITE_TABLE", "CRCON"))
    return quote_identifier(os.environ.get("TARGET_TABLE", os.environ.get("DB2_TARGET_TABLE", "CRCON")))


def target_key_column() -> str:
    if target_db() == "sqlite":
        return quote_identifier(os.environ.get("TARGET_SQLITE_KEY_COLUMN", "NODOSS"))
    return quote_identifier(os.environ.get("TARGET_KEY_COLUMN", os.environ.get("DB2_CONTRACT_COLUMN", "NODOSS")))


def build_update_sql(updates: list[dict[str, str]], placeholder: str = "?", table: str | None = None, key_column: str | None = None) -> str:
    assignments = ", ".join(f"{quote_identifier(item['db_column'])} = {placeholder}" for item in updates)
    return f"UPDATE {quote_identifier(table) if table else target_table()} SET {assignments} WHERE {quote_identifier(key_column) if key_column else target_key_column()} = {placeholder}"


def build_filled_update_preview(contract_number: str, updates: list[dict[str, str]], table: str | None = None, key_column: str | None = None) -> str:
    assignments = ", ".join(f"{quote_identifier(item['db_column'])} = {sql_literal(item['value'])}" for item in updates)
    return f"UPDATE {quote_identifier(table) if table else target_table()} SET {assignments} WHERE {quote_identifier(key_column) if key_column else target_key_column()} = {sql_literal(contract_number)}"


def update_target_contract(contract_number: str, updates: list[dict[str, str]], table: str | None = None, key_column: str | None = None) -> TargetResult:
    db = target_db()
    if db == "postgres":
        return update_postgres(contract_number, updates, table, key_column)
    if db in {"odbc", "db2", "ibmi"}:
        return update_odbc(contract_number, updates, table, key_column)
    return update_sqlite(contract_number, updates, table, key_column)


def update_sqlite(contract_number: str, updates: list[dict[str, str]], table: str | None = None, key_column: str | None = None) -> TargetResult:
    db_path = Path(os.environ.get("TARGET_SQLITE_PATH", "target.sqlite3"))
    sql = build_update_sql(updates, table=table, key_column=key_column)
    params = [item["value"] for item in updates] + [contract_number]
    try:
        with sqlite3.connect(db_path) as conn:
            cursor = conn.execute(sql, params)
            conn.commit()
        rows = cursor.rowcount or 0
        return TargetResult(ok=rows > 0, rows_affected=rows, error="" if rows > 0 else "SQLite a mis a jour 0 ligne.", sql=sql)
    except Exception as exc:
        return TargetResult(ok=False, error=str(exc), sql=sql)


def update_postgres(contract_number: str, updates: list[dict[str, str]], table: str | None = None, key_column: str | None = None) -> TargetResult:
    import psycopg

    sql = build_update_sql(updates, "%s", table=table, key_column=key_column)
    params = [item["value"] for item in updates] + [contract_number]
    dsn = os.environ.get("TARGET_DATABASE_URL") or os.environ.get("POSTGRES_URL")
    if not dsn:
        return TargetResult(ok=False, error="TARGET_DATABASE_URL n'est pas configure.", sql=sql)
    try:
        with psycopg.connect(dsn) as conn:
            with conn.cursor() as cursor:
                cursor.execute(sql, params)
                rows = cursor.rowcount or 0
        return TargetResult(ok=rows > 0, rows_affected=rows, error="" if rows > 0 else "Postgres a mis a jour 0 ligne.", sql=sql)
    except Exception as exc:
        return TargetResult(ok=False, error=str(exc), sql=sql)


def update_odbc(contract_number: str, updates: list[dict[str, str]], table: str | None = None, key_column: str | None = None) -> TargetResult:
    import pyodbc

    sql = build_update_sql(updates, table=table, key_column=key_column)
    params = [item["value"] for item in updates] + [contract_number]
    connection_string = os.environ.get("ODBC_CONNECTION_STRING") or os.environ.get("DB2_CONNECTION_STRING")
    autocommit = os.environ.get("ODBC_AUTOCOMMIT", "true").lower() in {"1", "true", "yes", "on"}
    if not connection_string:
        return TargetResult(ok=False, error="ODBC_CONNECTION_STRING n'est pas configure.", sql=sql)
    try:
        with pyodbc.connect(connection_string, autocommit=autocommit) as conn:
            cursor = conn.cursor()
            cursor.execute(sql, params)
            rows = cursor.rowcount if cursor.rowcount != -1 else 0
            if not autocommit:
                conn.commit()
        return TargetResult(ok=rows > 0, rows_affected=rows, error="" if rows > 0 else "ODBC a mis a jour 0 ligne.", sql=sql)
    except Exception as exc:
        return TargetResult(ok=False, error=str(exc), sql=sql)


def sql_literal(value: str) -> str:
    return "'" + value.replace("'", "''") + "'"
